framework/src/main/java/org/checkerframework/framework/util/ExpressionTreeToJavaExpressionVisitor.java

```diff
@@ -0,0 +1,1025 @@
+    setResolverField();
+    String idName = idTree.getName().toString();
+    // this and super logic
+    if (idName.equals("this") || idName.equals("super")) {
+      if (thisReference == null) {
+        throw new JavaExpressionParseExceptionUnchecked(
+            JavaExpressionParseException.construct(
+                idName, "\"" + idName + "\" is not allowed here"));
+      }
+      if (idName.equals("this")) {
+        return thisReference;
+      } else {
+        // super literal
+        TypeMirror superclass = TypesUtils.getSuperclass(enclosingType, types);
+        if (superclass == null) {
+          throw new JavaExpressionParseExceptionUnchecked(
+              JavaExpressionParseException.construct(
+                  "super", enclosingType + " has no superclass"));
+        }
+        return new SuperReference(superclass);
+      }
+    }
+
+    // Formal parameter, using "#2" syntax.
+    JavaExpression parameter = parseAsParameter(idName);
+    if (parameter != null) {
+      return parameter;
+    }
+
+    // Local variable or parameter.
+    // Attempt to match a local variable within the scope of the
+    // given path before attempting to match a field.
+    if (localVarPath != null) {
+      VariableElement varElem = resolver.findLocalVariableOrParameter(idName, localVarPath);
+      if (varElem != null) {
+        return new LocalVariable(varElem);
+      }
+    }
+
+    // Field access
+    JavaExpression fieldAccessReceiver;
+    if (thisReference != null) {
+      fieldAccessReceiver = thisReference;
+    } else {
+      fieldAccessReceiver = new ClassName(enclosingType);
+    }
+    FieldAccess fieldAccess = getIdentifierAsFieldAccess(fieldAccessReceiver, idName);
+    if (fieldAccess != null) {
+      return fieldAccess;
+    }
+
+    // Class name
+    if (localVarPath != null) {
+      Element classElem = resolver.findClass(idName, localVarPath);
+      TypeMirror classType = ElementUtils.getType(classElem);
+      if (classType != null) {
+        return new ClassName(classType);
+      }
+    }
+    ClassName classType = getIdentifierAsUnqualifiedClassName(idName);
+    if (classType != null) {
+      return classType;
+    }
+
+    // Err if a formal parameter name is used, instead of the "#2" syntax.
+    if (parameters != null) {
+      for (int i = 0; i < parameters.size(); i++) {
+        Element varElt = parameters.get(i).getElement();
+        if (varElt.getSimpleName().contentEquals(idName)) {
+          throw new JavaExpressionParseExceptionUnchecked(
+              JavaExpressionParseException.construct(
+                  idName,
+                  String.format(DependentTypesError.FORMAL_PARAM_NAME_STRING, i + 1, idName)));
+        }
+      }
+    }
+
+    throw new JavaExpressionParseExceptionUnchecked(
+        JavaExpressionParseException.construct(idName, "identifier not found"));
+  }
+
+  /**
+   * If {@code s} is a parameter expressed using the "_param_NN" syntax, then returns a
```

Comment:
Did the syntax change from `#NN`? Oh I see it is preprocessed to `_param_NN`.   A comment here would be useful.  Especially since the error message references the other syntax.